File: src/openclaw_predicate_provider/adapter.py
"""OpenClaw adapter scaffolding for guarded tool execution."""

from __future__ import annotations

from typing import Any, Awaitable, Callable

from .provider import GuardedProvider

ToolCallable = Callable[[dict[str, Any]], Awaitable[Any]]


class ToolAdapter:
    """Wrap tool handlers with Predicate guard checks."""

    def __init__(self, guard: GuardedProvider):
        self._guard = guard

    async def run(
        self,
        *,
        action: str,
        resource: str,
        args: dict[str, Any],
        context: dict[str, Any],
        execute: ToolCallable,
    ) -> Any:
        await self._guard.guard_or_raise(
            action=action,
            resource=resource,
            args=args,
            context=context,
        )
        return await execute(args)

    async def run_shell(
        self,
        *,
        args: dict[str, Any],
        context: dict[str, Any],
        execute: ToolCallable,
    ) -> Any:
        """Guard OpenClaw command execution."""
        resource = str(args.get("command", ""))
        return await self.run(
            action="shell.execute",
            resource=resource,
            args=args,
            context=context,
            execute=execute,
        )

    async def read_file(
        self,
        *,
        args: dict[str, Any],
        context: dict[str, Any],
        execute: ToolCallable,
    ) -> Any:
        """Guard OpenClaw file-read execution."""
        resource = str(args.get("path", ""))
        return await self.run(
            action="fs.read",
            resource=resource,
            args=args,
            context=context,
            execute=execute,
        )

    async def http_request(
        self,
        *,
        args: dict[str, Any],
        context: dict[str, Any],
        execute: ToolCallable,
    ) -> Any:
        """Guard OpenClaw outbound HTTP execution."""
        resource = str(args.get("url", ""))
        return await self.run(
            action="net.http",
            resource=resource,
            args=args,
            context=context,
            execute=execute,
        )

File: src/adapter.ts
import type { GuardedProvider } from "./provider.js";

export interface ToolRunArgs {
  args: Record<string, unknown>;
  context: Record<string, unknown>;
  execute: (args: Record<string, unknown>) => Promise<unknown>;
}

export class ToolAdapter {
  constructor(private readonly guard: Pick<GuardedProvider, "guardOrThrow">) {}

  async run(params: ToolRunArgs & { action: string; resource: string }) {
    await this.guard.guardOrThrow({
      action: params.action,
      resource: params.resource,
      args: params.args,
      context: params.context,
    });
    return params.execute(params.args);
  }

  async runShell(params: ToolRunArgs) {
    return this.run({
      ...params,
      action: "shell.execute",
      resource: String(params.args.command ?? ""),
    });
  }

  async readFile(params: ToolRunArgs) {
    return this.run({
      ...params,
      action: "fs.read",
      resource: String(params.args.path ?? ""),
    });
  }

  async httpRequest(params: ToolRunArgs) {
    return this.run({
      ...params,
      action: "net.http",
      resource: String(params.args.url ?? ""),
    });
  }
}
